### PR TITLE
#8 [4/4]: Return model object

### PR DESCRIPTION
**Pull request by zsusswein**
Return the fitted model object with class `RtGam`. The package now does something 🥳 

As part of returning the object, automatically extract some diagnostics about fitting and issue warnings if clearly unsatisfactory. Some of these choices are a bit arbitrary, but I don't want to let the perfect to be the enemy of the good. 

Most of the new functionality is in private functions. There are a few public functions and I've made an effort to have some clear-ish documentation there.

There's on NOTE in R CMD check from adding `{withr}` in `imports` without using it in the main package. I think it's fine there though because it's going to get more use in the next round of PRs and it's used in the examples.

### CODE REVIEW

**Comment by seabbs** [comment]
> The package now does something 🥳

Wow

**Comment by seabbs** [comment]
> Some of these choices are a bit arbitrary, but I don't want to let the perfect to be the enemy of the good.

For the things you feel that this is the case for can you make some issues to investigate etc vs us discussing here?

**Comment by seabbs** [comment]
> There's on NOTE in R CMD check from adding {withr} in imports without using it in the main package. I think it's fine there though because it's going to get more use in the next round of PRs and it's used in the examples.

Again I would make an issue to track this

File: R/RtGam.R [hunk 1]
```diff
@@ -37,6 +37,10 @@
 #'   [mgcv::bam()] converges more quickly but introduces some additional
 #'   numerical error. Note that the `bam` backend uses the `discrete = TRUE`
 #'   option for an additional speedup. See [mgcv::bam()] for more information.
+#' @param warn_for_diagnostic_failure Should warnings be issued for
```
Comment (seabbs) [inline R/RtGam.R:40]: another option (for its own issue) is to have a package wide verbose mode you control via `options()`.

File: R/RtGam.R [hunk 2]
```diff
@@ -37,6 +37,10 @@
...
+#'   automatically identified diagnostic issues? Defaults to true. A list of
```
Comment (seabbs) [inline R/RtGam.R:41]:
```suggestion
#'   automatically identified diagnostic issues? Defaults to TRUE. A list of
```

I am not sure that this is good practice but is what I usually do.

File: R/RtGam.R [hunk 3]
```diff
@@ -48,15 +52,23 @@
 #' @return Stub function: NULL
 #' @export
 #' @examples
-#' cases <- c(1L, 2L, 3L)
-#' reference_date <- as.Date(c("2023-01-01", "2023-01-02", "2023-01-03"))
-#' mod <- RtGam::RtGam(cases, reference_date)
+#' withr::with_seed(12345, {
```
Comment (seabbs) [inline R/RtGam.R:55]: do you really need this?

File: R/RtGam.R [hunk 4]
```diff
@@ -48,15 +52,23 @@
...
+#'   cases <- rpois(20, 10)
+#' })
+#' reference_date <- seq.Date(
+#'   from = as.Date("2023-01-01"),
+#'   length.out = 20,
+#'   by = "day"
+#' )
+#' fit <- RtGam::RtGam(cases, reference_date)
```
Comment (seabbs) [inline R/RtGam.R:63]: kind of unusual to use package::function in your own package examples?

File: R/RtGam.R [hunk 5]
```diff
@@ -83,8 +95,21 @@ RtGam <- function(cases,
       ...
     )
   )
+  diagnostics <- calculate_diagnostics(fit)
+  if (warn_for_diagnostic_failure) {
+    issue_diagnostic_warnings(diagnostics)
+  }
 
-  invisible(NULL)
+  format_for_return(
```
Comment (seabbs) [inline R/RtGam.R:103]: is this really a class constructor? If yes what about calling it that (i.e new_RtGam) just for R standard following

File: R/RtGam.R [hunk 6]
```diff
@@ -83,8 +95,21 @@ RtGam <- function(cases,
...
+    fit = fit,
+    df = df,
+    group = group,
+    k = k,
+    m = m,
+    backend = backend,
+    formula = formula,
+    diagnostics = diagnostics
```
Comment (seabbs) [inline R/RtGam.R:111]: I don't feel that strongly but do you feel strongly that there is value to storing the diagnostics vs just having users call `check_diagnostics`?

File: R/diagnostics.R [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+#'   cases <- rpois(20, 10)
+#' })
+#' reference_date <- seq.Date(
+#'   from = as.Date("2023-01-01"),
+#'   length.out = 20,
+#'   by = "day"
+#' )
+#' fit <- RtGam::RtGam(cases, reference_date)
+#' check_diagnostics(fit)
+check_diagnostics <- function(fit, warn_for_diagnostic_failure = TRUE) {
```
Comment (seabbs) [inline R/diagnostics.R:46]: what about having an internal version of this that works with fit (i.e _check_diagnostics) and that you use here and in the main function?

File: R/RtGam.R (hunk 3 above)
Comment (zsusswein) [reply R/RtGam.R:55]: I don't _really_. But I'm throwing a warning whenever the k-index is below one. Under the null that happens 5% of the time. So fitting it to random white noise will randomly throw warnings which I don't love.

File: tests/testthat/test-RtGam.R [hunk 8]
```diff
@@ -1,10 +1,26 @@
 test_that("RtGam parses inputs successfully", {
   timesteps <- 20
-  cases <- 1:timesteps
+  withr::with_seed(12345, {
```
Comment (seabbs) [inline tests/testthat/test-RtGam.R:3]: as it stands i would have this in suggests

**Review by seabbs — APPROVED** [review]
Overall this looks good. Some very minor points that you can push back on about style but functionality wise it looks good.

**Review by seabbs — APPROVED** [review]
Overall this looks good. Some very minor points that you can push back on about style but functionality wise it looks good.

File: R/RtGam.R (hunk 6 above)
Comment (zsusswein) [reply R/RtGam.R:111]: I do. I think we want the diagnostics in the returned object and `check_diagnostics()` is ancillary. I was debating whether to have `check_diagnostics()` at all.

The key point here is that I think we want to do automated checks as a first class citizen. They're a primary concern and part of any model run. Allowing the users to run them again is nice but ancillary. And if we're going to generate the diagnostics, we should at least store them.

File: R/RtGam.R (hunk 3 above)
Comment (seabbs) [reply R/RtGam.R:55]: yeah that makes sense though that high a false positive rate might get pretty annoying pretty fast

File: R/diagnostics.R (hunk 7 above)
Comment (zsusswein) [reply R/diagnostics.R:46]: Isn't that what I do with `calculate_diagnostics()`?

File: R/RtGam.R (hunk 3 above)
Comment (zsusswein) [reply R/RtGam.R:55]: The joys of NHST

File: R/diagnostics.R (hunk 7 above)
Comment (seabbs) [reply R/diagnostics.R:46]: you do but you duplicate the if else check and the call to issues

File: R/RtGam.R (hunk 6 above)
Comment (seabbs) [reply R/RtGam.R:111]: > Allowing the users to run them again is nice but ancillary. 

Is it nice? As in when would they do this vs look at the diagnostics in the fit output (i.e your point)

File: R/diagnostics.R (hunk 7 above)
Comment (zsusswein) [reply R/diagnostics.R:46]: Oh duh. I'll just call the public func internally.

File: R/RtGam.R (hunk 6 above)
Comment (seabbs) [reply R/RtGam.R:111]: you could have it so that this just accesses the diagnostics in the fit object for people

File: R/RtGam.R (hunk 1 above)
Comment (zsusswein) [reply R/RtGam.R:40]: #28

File: R/RtGam.R (hunk 5 above)
Comment (zsusswein) [reply R/RtGam.R:103]: I don't think it is the R standard for model fitting functions? e..g., `lm()`, `glm()`, `gam()` all return an eponymously classed object. They're class constructors but the standard for unifunctional packages is to just construct with a function of the same name. 

I could have sworn Hadley says something along these lines in R packages 2e but I couldn't find it with a quick cmd+F

File: tests/testthat/test-RtGam.R (hunk 8 above)
Comment (zsusswein) [reply tests/testthat/test-RtGam.R:3]: https://github.com/CDCgov/cfa-gam-rt/pull/22/commits/a841218d8dd92a6beb12c93e753e311e6789ad46

File: R/diagnostics.R (hunk 7 above)
Comment (zsusswein) [reply R/diagnostics.R:46]: https://github.com/CDCgov/cfa-gam-rt/pull/22/commits/1f9702e1f5942829f074e37939fe0256125f40b2

File: R/RtGam.R (hunk 6 above)
Comment (zsusswein) [reply R/RtGam.R:108]: https://github.com/CDCgov/cfa-gam-rt/pull/22/commits/1f9702e1f5942829f074e37939fe0256125f40b2

File: R/RtGam.R (hunk 4 above)
Comment (zsusswein) [reply R/RtGam.R]: https://github.com/CDCgov/cfa-gam-rt/pull/22/commits/a8fa503c606ec9facb3814ed0de4f34284d6d995

File: R/RtGam.R (hunk 5 above)
Comment (zsusswein) [reply R/RtGam.R]: https://github.com/CDCgov/cfa-gam-rt/pull/22/commits/c5f808cdfab3b01b234351038cd70f216e4e63e3

**Comment by zsusswein** [comment]
> > Some of these choices are a bit arbitrary, but I don't want to let the perfect to be the enemy of the good.
> 
> For the things you feel that this is the case for can you make some issues to investigate etc vs us discussing here?

#29

**Review by seabbs — APPROVED** [review]
LGTM

**Pull request merged**